### PR TITLE
Use elasticsearch to get group annotation count

### PR DESCRIPTION
**Pull request by hmstepanek**
Previously the count was obtained from a db query. The problem with
this was two fold:
 1. Shared was being queried on even though it wasn't indexed
 2. When searching for groups such as Public, where the group has a
    large portion of the annotation in the table, the db was optimizing
    to search the table directly rather than use the index due to the
    overhead of indexing lookup being larger than a full table scan.

Elasticsearch performs much better for aggregation queries with 
complex where clauses and it also has built in caching. Although
looking at the Public group on activity pages doesn't happen that 
often, when it does, it takes ~3.5 s on average of query time to perform 
this lookup. ES ends up taking ~.01 s.

This is a partial fix for https://github.com/hypothesis/product-backlog/issues/812.

Since the count is now coming from elasticsearch with the default query
filters, this count now takes into account other things such as; nipsa,
permissions, moderation, etc. While the count previously considered
"annotations" to include replies, since this is used on activity pages
only, it was agreed by the team that it makes more sense for this to
not include replies and exactly match the "Matching Annotations" count.

This decision of not including replies also has the advantage that if
the query run in the search box doesn't include any other search
criteria we can re-use that total and we don't have to run this query
at all.

### CODE REVIEW

File: tests/h/views/admin/users_test.py [hunk 1]
```diff
@@ -240,8 +240,8 @@ def user_service(pyramid_config, db_session):
 
 
 @pytest.fixture
-def annotation_stats_service(pyramid_config, db_session):
-    service = Mock(spec_set=AnnotationStatsService(session=db_session))
+def annotation_stats_service(pyramid_config, pyramid_request):
+    service = Mock(spec_set=AnnotationStatsService(request=pyramid_request))
```
Comment (seanh) [inline tests/h/views/admin/users_test.py]: I know you're just replacing an existing `Mock` with a `Mock` but changing this to something like `mock.create_autospec(AnnotationStatsService, instance=True, spec_set=True, request=pyramid_request)` would be better. You'd get a better-specced mock and also these tests would not be instantiating `AnnotationStatsService`

**Comment by seanh** [comment]
Another approach to this might be to pre-compute group stats such as the total number of annotations in the group. Store the group's total number of annotations in a column on the group table in the DB. When an annotation is _created_ in a group, just increment this stored number by 1.

File: h/services/annotation_stats.py [hunk 2]
```diff
@@ -38,13 +41,18 @@ def group_annotation_count(self, pubid):
         """
         Return the count of shared annotations for this group.
         """
+        search = Search(self.request, stats=self.request.stats)
+        search.clear()
+        search.append_modifier(query.Limiter())
+        search.append_modifier(query.AuthFilter(self.request))
+        search.append_modifier(query.DeletedFilter())
```
Comment (robertknight) [inline h/services/annotation_stats.py]: When I view http://localhost:5000/groups/__world__/public on my system, I see a much higher number for the annotation count than I do on master. Dumping the executed search queries using `search.to_dict` in `h.search.Search._search`, I see the following query executed:

```json
{
  "query": {
    "bool": {
      "filter": [
        {
          "bool": {
            "must_not": [
              {
                "exists": {
                  "field": "deleted"
                }
              }
            ]
          }
        },
        {
          "bool": {
            "should": [
              {
                "term": {
                  "shared": true
                }
              },
              {
                "term": {
                  "user_raw": "acct:robert@localhost"
                }
              }
            ]
          }
        }
      ]
    }
  },
  "sort": [
    {
      "updated": {
        "order": "desc",
        "unmapped_type": "boolean"
      }
    }
  ],
  "from": 0,
  "size": 0,
  "_source": false
}
```

I can see a few problems with this:

1. Annotations are not actually filtered by group.
2. Replies are not filtered out
3. Moderated and nipsa'd annotations are not filtered out

If I delete the `search.clear()` and `append_modifier` lines then the returned count is consistent with master.

Comment (hmstepanek) [reply h/services/annotation_stats.py]: The way this query was written before on the db, none of those things were true before. Note that more annotations are shown in the group count from the db on the right than there are in the annotation count from search. 
image

That's not to say we can't change it. Personally I think it's a bit confusing as I would expect those numbers to match but I was trying to keep as close as possible to the existing behavior. The only intended difference is the logged in user's private annotations are included in the count.

I think the one issue here as Sean pointed out-I durpely overlooked attaching the GroupFilter and the tests didn't catch it cause they don't create another group.

**Comment by hmstepanek** [comment]
I do think there is an optimization we can make in activity pages that if there is no extra query params then we can re-use the search count returned from that without running this query again (assuming that query matches this query-right now they don't but we could change this).

File: tests/h/services/annotation_stats_test.py [hunk 3]
```diff
@@ -69,36 +72,93 @@ def test_user_annotation_counts_ignores_deleted_annotations(self, svc, factories
         assert results['group'] == 0
         assert results['total'] == 0
 
-    def test_annotation_count_returns_count_of_shared_annotations_for_group(self, svc, factories):
+    def test_annotation_count_returns_count_of_shared_annotations_for_group(self, svc, Annotation):
         pubid = 'abc123'
-        for i in range(3):
-            factories.Annotation(groupid=pubid, shared=True)
+        Annotation(groupid="wrong_grp", shared=True)
         for i in range(2):
```
Comment (hmstepanek) [inline tests/h/services/annotation_stats_test.py]: I felt we were creating an unnecessary amount of annotations here so I decreased the numbers a little.

File: h/views/activity.py [hunk 4]
```diff
@@ -205,6 +205,19 @@ def user_annotation_count(aggregation, userid):
...
         return result
 
+    def _get_total_annotations_in_group(self, result, q, request):
+        """
+        Get number of annotations in group.
+
+        If the search result already has this number don't run a query, just re-use it.
+        """
+        group_annotation_count = result['search_results'].total
+        if len(q) > 1:
```
Comment (robertknight) [inline h/views/activity.py]: This is a good optimization to use, although I think the test for whether it can be safely used is perhaps a little brittle. A more explicit test for whether the search result is for a single group would be preferable.

File: tests/h/services/annotation_stats_test.py [hunk 5]
```diff
@@ -92,13 +109,32 @@ def test_returns_service(self):
...
+
+
+@pytest.fixture
+def svc(pyramid_request):
+    return AnnotationStatsService(request=pyramid_request)
+
+
+@pytest.fixture
+def pyramid_request(pyramid_request):
+    pyramid_request.es = mock.Mock()
```
Comment (robertknight) [inline tests/h/services/annotation_stats_test.py]: `pyramid_request.es` doesn't need to be mocked any more.

File: tests/h/views/activity_test.py [hunk 6]
```diff
@@ -573,10 +575,36 @@ def test_search_passes_the_group_annotation_count_to_the_template(self,
                                                                       factories,
                                                                       test_group,
                                                                       test_user,
+                                                                      query,
+                                                                      annotation_stats_service,
                                                                       pyramid_request):
+        query.extract.return_value = MultiDict({"user": test_user, "group": test_group})
         result = controller.search()['stats']
+        annotation_stats_service.group_annotation_count.assert_called_with(test_group.pubid)
         assert result['annotation_count'] == 5
```
Comment (robertknight) [inline tests/h/views/activity_test.py]: It wasn't immediately clear when reading this that we are verifying that the `annotation_count` template value is the same as the return value from `annotation_stats_service.group_annotation_count`.

File: tests/h/views/admin/users_test.py [hunk 7]
```diff
@@ -233,22 +233,35 @@ def models(patch):
 
 @pytest.fixture
 def user_service(pyramid_config, db_session):
-    service = Mock(spec_set=UserService(default_authority='example.com',
-                                        session=db_session))
+    service = mock.create_autospec(
```
Comment (robertknight) [inline tests/h/views/admin/users_test.py]: These changes look fine, but they appear to be unrelated to the purpose of the PR? It's always fine to submit small cleanup PRs separately - whether they relate to the code you are currently working on or not.

**Comment by robertknight** [comment]
> Another approach to this might be to pre-compute group stats such as the total number of annotations in the group. 

We could do that, but then we have fun cache invalidation challenges to solve. In the end it turned out there were other advantages to using search - such as being able to respect moderation and privacy status without having to write any extra code.

**Comment by hmstepanek** [comment]
That multisearch sounds interesting! -should make a note of that for when we refactor activity pages I think.
